fog/sample-paykit/src/client.rs

```diff
@@ -362,6 +360,280 @@ impl Client {
+        let input_amount = input.amount;
+        assert!(
+            offered.token_id == input_amount.token_id,
+            "get_transaction_inputs post condition failed"
+        );
+        assert!(
+            offered.value <= input_amount.value,
+            "get_transaction_inputs post condition failed"
+        );
+        let change = Amount::new(input_amount.value - offered.value, offered.token_id);
+
+        let tombstone_block = self.compute_tombstone_block()?;
+
+        let block_version = BlockVersion::try_from(self.tx_data.get_latest_block_version())?;
+
+        // Make fog resolver
+        let fog_uris = self
+            .account_key
+            .fog_report_url()
+            .map(FogUri::from_str)
+            .transpose()?;
+        let fog_responses = self
+            .fog_report_conn
+            .fetch_fog_reports(fog_uris.into_iter())?;
+        let fog_resolver = FogResolver::new(fog_responses, &self.fog_verifier)?;
+
+        let (ring, membership_proofs): (Vec<TxOut>, Vec<TxOutMembershipProof>) =
+            ring.into_iter().unzip();
+        let input_credentials = input_credentials_helper(
+            input,
+            input_proof,
+            ring,
+            membership_proofs,
+            &self.account_key,
+        )?;
+
+        // TODO: Use the RTHMemoBuilder?
+        let mut sci_builder = SignedContingentInputBuilder::new(
+            block_version,
+            input_credentials,
+            fog_resolver,
+            EmptyMemoBuilder::default(),
+        )?;
+
+        sci_builder
+            .add_required_output(requested, &self.account_key.default_subaddress(), rng)
+            .map_err(Error::AddOutput)?;
+
+        let change_destination = ReservedDestination::from(&self.account_key);
+
+        sci_builder
+            .add_required_change_output(change, &change_destination, rng)
+            .map_err(|err| {
+                log::error!(self.logger, "Could not add change due to {:?}", err);
+                Error::AddOutput(err)
+            })?;
+
+        sci_builder.set_tombstone_block(tombstone_block);
+
+        Ok(sci_builder.build(rng)?)
+    }
+
+    /// Builds a transaction that fulfills a swap request, sending all excess
+    /// funds to ourselves and paying fee.
+    ///
+    /// # Arguments
+    /// * `sci` - The swap request we are fulfilling.
+    /// * `fee` - The transaction fee to use.
+    /// * `rng` - Randomness.
+    pub fn build_swap_transaction<T: RngCore + CryptoRng>(
+        &mut self,
+        mut sci: SignedContingentInput,
+        fee: Amount,
+        rng: &mut T,
+    ) -> Result<Tx> {
+        mc_common::trace_time!(self.logger, "MobileCoinClient.build_swap_transaction");
+
+        // Validate the sci
+        sci.validate()?;
+
+        // Check if its key image alreay landed
+        // Note: An actual fog wallet may not want to do this part, because it adds to
+        // latency, but if you have a local ledger copy you should definitely do this
```

Comment:
Nit: Might be worth noting that this is still racy - i.e. if this test passes it is no guarantee that the SCI won't expire between now and when it is submitted.